File: python/src/procedural_memory.py
from dataclasses import dataclass
import math

from generated.grpc.actr import NeuroAction, ProceduralCondition
from generated.grpc.actr.services import ProceduralMemoryServiceBase, GetAllConditionsResponse, SelectRuleRequest, \
    LearnUtilityRequest

import random
from betterproto.lib.google.protobuf import Empty


@dataclass
class Rule:
    id: str
    condition: ProceduralCondition
    action: NeuroAction
    utility: float


class ProceduralMemory(ProceduralMemoryServiceBase):
    def __init__(self, temperature: float = 0.5, learning_rate: float = 0.1):
        self.rules: dict[str, Rule] = {}
        self.temperature = temperature
        self.lr = learning_rate

    async def get_all_conditions(self, betterproto_lib_google_protobuf_empty) -> GetAllConditionsResponse:
        _ = betterproto_lib_google_protobuf_empty

        return GetAllConditionsResponse(conditions=[r.condition for r in self.rules.values()])

    async def select_rule(self, select_rule_request: SelectRuleRequest) -> NeuroAction:
        _ = select_rule_request

        applicable = [r for r in self.rules.values() if (r.id in select_rule_request.satisfied_rule_ids)]
        if not applicable:
            raise ValueError("No applicable rule found.")

        utilities = [r.utility for r in applicable]
        max_u = max(utilities)
        exp_utils = [math.exp((u - max_u) / self.temperature) for u in utilities]
        sum_exp = sum(exp_utils)

        probs = [e / sum_exp for e in exp_utils]
        rule = random.choices(applicable, weights=probs, k=1)[0]

        return rule.action

    async def learn_utility(self, learn_utility_request: LearnUtilityRequest) -> Empty:
        rule_id = learn_utility_request.rule_id

        if rule_id in self.rules:
            rule = self.rules[rule_id]
            rule.utility += self.lr * (learn_utility_request.reward - rule.utility)

        return Empty()

File: Dotnet/src/Harness.Core/ProceduralMemory.cs
using Harness.Abstractions;
using Harness.Abstractions.Actr;
using Harness.Core.Configuration;
using Harness.Core.Rules;
using Harness.Shared.Observability;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace Harness.Core;

public class ProceduralMemory : IProceduralMemory, IProvideLogger
{
    private sealed class RuleState
    {
        public required string Id { get; init; }
        public required ProceduralCondition Condition { get; init; }
        public required NeuroAction Action { get; init; }
        public required double Utility { get; set; }
    }

    private readonly Dictionary<string, RuleState> _rules;
    private readonly ILogger<ProceduralMemory> _logger;
    private readonly double _temperature;
    private readonly double _learningRate;
    private readonly Random _random;
    private string? _lastRuleId;

    public ProceduralMemory(
        RulesLoader rulesLoader,
        IOptions<ProceduralMemoryOptions> options,
        IClock clock,
        ILogger<ProceduralMemory> logger)
    {
        _logger = logger;
        _temperature = options.Value.Temperature;
        _learningRate = options.Value.LearningRate;
        _random = new Random(options.Value.RandomSeed);
        _rules = rulesLoader.LoadRules().ToDictionary(
            pair => pair.Key,
            pair => new RuleState
            {
                Id = pair.Value.Id,
                Condition = pair.Value.Condition.Clone(),
                Action = pair.Value.Action.Clone(),
                Utility = pair.Value.Utility
            },
            StringComparer.Ordinal);

        clock.OnTickAsync += (reward, ct) =>
        {
            if (!reward.Training)
            {
                return Task.CompletedTask;
            }

            return _lastRuleId == null ? throw new InvalidOperationException() : LearnUtilityAsync(reward.Reward, ct);
        };
    }

    public ILogger Logger => _logger;

    [TraceSpan]
    public IReadOnlyList<ProceduralCondition> GetAllConditions()
    {
        return _rules.Values
            .Select(rule => rule.Condition.Clone())
            .ToList();
    }

    [TraceSpan]
    public NeuroAction SelectRule(IReadOnlyList<string> satisfiedRuleIds)
    {
        var applicableRules = _rules.Values
            .Where(rule => satisfiedRuleIds.Contains(rule.Id, StringComparer.Ordinal))
            .ToList();

        if (applicableRules.Count == 0)
        {
            throw new InvalidOperationException(
                $"No applicable rule found for satisfied_rule_ids=[{string.Join(", ", satisfiedRuleIds)}].");
        }

        RuleState selectedRule;
        if (_temperature <= 0)
        {
            selectedRule = applicableRules
                .OrderByDescending(rule => rule.Utility)
                .ThenBy(rule => rule.Id, StringComparer.Ordinal)
                .First();

            TracingModel.AddEvent(
                TracingModel.Events.RuleSelected,
                new[]
                {
                    new KeyValuePair<string, object?>(
                        TracingModel.Tags.RuleId,
                        selectedRule.Id),
                    new KeyValuePair<string, object?>(
                        TracingModel.Tags.RuleCandidateCount,
                        applicableRules.Count),
                    new KeyValuePair<string, object?>(
                        TracingModel.Tags.RuleSelectionMode,
                        "deterministic")
                });
        }
        else
        {
            var maxUtility = applicableRules.Max(rule => rule.Utility);
            var weightedRules = applicableRules
                .Select(rule => new
                {
                    Rule = rule,
                    Weight = Math.Exp((rule.Utility - maxUtility) / _temperature)
                })
                .ToList();

            var weightSum = weightedRules.Sum(item => item.Weight);
            var sample = _random.NextDouble() * weightSum;
            var cumulative = 0.0;
            selectedRule = weightedRules[^1].Rule;

            foreach (var weightedRule in weightedRules)
            {
                cumulative += weightedRule.Weight;
                if (sample <= cumulative)
                {
                    selectedRule = weightedRule.Rule;
                    break;
                }
            }

            TracingModel.AddEvent(
                TracingModel.Events.RuleSelected,
                new[]
                {
                    new KeyValuePair<string, object?>(
                        TracingModel.Tags.RuleId,
                        selectedRule.Id),
                    new KeyValuePair<string, object?>(
                        TracingModel.Tags.RuleCandidateCount,
                        applicableRules.Count),
                    new KeyValuePair<string, object?>(
                        TracingModel.Tags.RuleSelectionMode,
                        "stochastic")
                });
        }

        _lastRuleId = selectedRule.Id;
        return selectedRule.Action.Clone();
    }

    [TraceSpan]
    private async Task LearnUtilityAsync(float reward, CancellationToken cancellationToken = default)
    {
        cancellationToken.ThrowIfCancellationRequested();

        if (_lastRuleId is null || !_rules.TryGetValue(_lastRuleId, out var rule))
        {
            return;
        }

        rule.Utility += _learningRate * (reward - rule.Utility);
        await Task.CompletedTask;
    }
}
